unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -352,6 +351,180 @@ public int indexOf(UTF8String v, int start) {
+  }
+
+  private int lastIndexOfInByte(UTF8String v, int fromIndexInByte) {
+    if (numBytes == 0) {
+      return -1;
+    }
+    do {
+      int startByteIndex = fromIndexInByte - v.numBytes + 1;
+      if (startByteIndex < 0 ) {
+        return -1;
+      }
+      if (ByteArrayMethods.arrayEquals(
+              base, offset + startByteIndex, v.base, v.offset, v.numBytes)) {
+        int count = 0; // count from right most to the match end in byte.
+        while (startByteIndex >= 0) {
+          count++;
+          startByteIndex = firstOfCurrentCodePoint(startByteIndex) - 1;
+        }
+        return count - 1;
+      }
+      fromIndexInByte  = firstOfCurrentCodePoint(fromIndexInByte) - 1;
+    } while (fromIndexInByte >= 0);
+    return -1;
+  }
+
+  /**
+   * Finds the n-th last index within a String.
+   * This method uses {@link String#lastIndexOf(String)}.</p>
+   *
+   * @param searchStr  the String to find, may be null
+   * @param ordinal  the n-th last <code>searchStr</code> to find
+   * @return the n-th last index of the search String,
+   *  <code>-1</code> if no match or <code>null</code> string input
+   */
+  protected int lastOrdinalIndexOf(
+          UTF8String searchStr,
+          int ordinal) {
+    return doOrdinalIndexOf(searchStr, ordinal, true);
+  }
+
+  /**
+   * Finds the n-th index within a String, handling <code>null</code>.
+   * A <code>null</code> String will return <code>-1</code>
+   *
+   * @param searchStr  the String to find, may be null
+   * @param ordinal  the n-th <code>searchStr</code> to find
+   * @return the n-th index of the search String,
+   *  <code>-1</code> if no match or <code>null</code> string input
+   */
+  protected int ordinalIndexOf(
+          UTF8String searchStr,
+          int ordinal) {
+    return doOrdinalIndexOf(searchStr, ordinal, false);
+  }
+
+  private int doOrdinalIndexOf(
+          UTF8String searchStr,
+          int ordinal,
+          boolean lastIndex) {
+    if (ordinal <= 0) {
+      return -1;
+    }
+    if (searchStr.numBytes == 0) {
+      return lastIndex ? numChars() : 0;
+    }
+    int found = 0;
+    int index = lastIndex ? numBytes : -1;
+    do {
+      if (lastIndex) {
+        index = lastIndexOfInByte(searchStr, index - 1);
+      } else {
+        index = indexOf(searchStr, index + 1);
+      }
+      if (index < 0) {
+        return index;
+      }
+      found += 1;
+    } while (found < ordinal);
+    return index;
+  }
+  /**
+   * Returns the substring from string str before count occurrences of the delimiter delim.
+   * If count is positive, everything the left of the final delimiter (counting from left) is
+   * returned. If count is negative, every to the right of the final delimiter (counting from the
+   * right) is returned. substring_index performs a case-sensitive match when searching for delim.
+   */
+  public UTF8String subStringIndex(UTF8String delim, int count) {
+    if (delim == null) {
```

Comment:
We dont' need to check the null value here, as it's done in the expression side.